src/lib/support/SerializableIntegerSet.h

```diff
@@ -0,0 +1,150 @@
+ *      capacity. The data structure supports serialize/deserialize operations,
+ *      where, serialize converts the set to a base64 string, and deserializing
+ *      the base64 string reconstructs the original set.
+ *
+ *      The data is stored such that serialized data can be deserialized correctly
+ *      on different machine architectures.
+ *
+ */
+
+#pragma once
+
+#include <support/Base64.h>
+#include <support/CodeUtils.h>
+
+namespace chip {
+
+class SerializableU64SetBase
+{
+
+public:
+    SerializableU64SetBase(uint64_t * data, uint16_t capacity, uint64_t emptyValue) :
+        mData(data), mCapacity(capacity), mEmptyValue(emptyValue), mNextAvailable(0)
+    {
+        for (uint16_t i = 0; i < capacity; i++)
+        {
+            data[i] = emptyValue;
+        }
+    }
+
+    /**
+     * @brief
+     *   Serialize the sparse array into a base 64 encoded string.
+     *   Only the values till mNextAvailable index are encoded.
+     *   The empty indexes between 0, and mNextAvailable, are also
+     *   encoded.
+     *
+     * @param[in] buf Buffer where serialized string is written
+     * @param[in] buflen Length of buf
+     * @return pointer to buf, or nullptr in case of error
+     */
+    const char * SerializeBase64(char * buf, uint16_t & buflen);
+
+    /**
+     * @brief
+     *   Deserialize a base64 encoded string into the sparse array.
+     *   The mNextAvailable index is calculated based on how many
+     *   values are in the deserialized array.
+     *
+     * @param[in] serialized Serialized buffer
+     * @param[in] buflen Length of buffer
+     * @return CHIP_NO_ERROR in case of success, or the error code
+     */
+    CHIP_ERROR DeserializeBase64(const char * serialized, uint16_t buflen);
+
+    /**
+     * @brief
+     *   Get the length of string if the array is serialized.
+     */
+    uint16_t SerializedSize() { return static_cast<uint16_t>(BASE64_ENCODED_LEN(sizeof(uint64_t) * mNextAvailable)); }
+
+    /**
+     * @brief
+     *   Get the maximum length of string if the array were full and serialized.
+     */
+    uint16_t MaxSerializedSize() { return static_cast<uint16_t>(BASE64_ENCODED_LEN(sizeof(uint64_t) * mCapacity)); }
+
+    /**
+     * @brief
+     *   Check if the value is in the array.
+     *
+     * @param[in] value Value to find
+     * @return True, if it's prsent in the array.
+     */
+    bool Contains(uint64_t value) { return Find(value) != mCapacity; }
+
+    /**
+     * @brief
+     *   Insert the value in the array. If the value is duplicate, it
+     *   won't be inserted.
+     *
+     * @return CHIP_NO_ERROR in case of success, or the error code
+     */
+    CHIP_ERROR Insert(uint64_t value);
+
+    /**
+     * @brief
+     *   Delete the value from the array.
+     */
+    void Remove(uint64_t value);
+
+private:
+    uint64_t * const mData;
+    const uint16_t mCapacity;
+    const uint64_t mEmptyValue;
```

Comment:
static constespr for this one ... I don't think we ever want to change it.

Reply:
The current code allows user to choose a specific empty value.
```
template <uint16_t kCapacity, uint64_t kEmptyValue = 0>
class SerializableU64Set : public SerializableU64SetBase
```

I think it could be useful in future.